lib/src/db/models.dart

```diff
@@ -10,12 +10,12 @@ part of gcloud.db;
  * The [Key] can be incomplete if it's id is `null`. In this case the id will
  * be automatically allocated and set at commit time.
  */
-class Key {
+class Key<T> {
   // Either KeyImpl or PartitionImpl
   final Object _parent;
 
   final Type type;
-  final Object id;
+  final T id;
 
   Key(Key parent, this.type, this.id) : _parent = parent {
```

Comment:
Do we still have a need to have `type` in the constructor. We can simply have a getter for backwards compatibility:

```dart
  Type get type => T;
```

right?

Reply:
Oh yes. Sounds good.

Reply:
Ah actually, this `type` represents the type of the entity that is stored with this key if I'm not mistaken, whereas `T` is the type of the `id`.

Reply:
Oh yes, I was reading this entirely wrong. I thought the new `T` is the model type, e.g. `Key<Person>`, which would allow us to have more type-safe code by knowing which model the key is referencing.

Reply:
I think that both would be really beneficial to avoid type mismatches. Given that this is a database library, where I think most people will agree that safety > convenience, I’d go for both.

The library itself will already provide the correctly typed keys, and we could add a convenience function to get a key from a model. 

Maybe there is a good solution for parent keys as well?